chapters/categorical-predictors.qmd

```diff
@@ -0,0 +1,718 @@
+Since there would be no influencer column in the data, all of the indicator columns would be zero when an influencer was observed in new data being predicted. However, from @tbl-ref-cell-effects we know that such a pattern would map to the reference level (which is the `"contract"` customer in our data). The new levels would be mapped to the reference level for this parameterization. This would occur more by happenstance and would not be a deliberate strategy. 
+
+Traditional indicator encoding methods should be avoided if new predictor categories are anticipated. Instead, the techniques used in the next section (effect encodings or feature hashing) will likely be more effective at dealing with this situation. 
+
+## Predictors with Many Values {#sec-high-cardinality}
+
+Some predictors have high cardinality; they can take many possible values relative to the number of data points. For example, in the entirety of the hotel rate data, there are `r length(unique(hotel_rate_train$agent))` unique agents (see @fig-agents). Should we create so many indicator variables for this predictor? In the data, `r hotel_rate_train %>% count(agent) %>% filter(n <= 5) %>% nrow()`  agents had fewer than six bookings. This means there will be many sparse columns with almost all zero entries. Additionally, `r sum(!(unique(hotel_rate_train$agent) %in% unique(hotel_rate_test$agent)))` of the agents have no data in the training set; they only occur in the test set. Their indicator columns would contain all zeros. 
+
+```{r}
+#| label: fig-agents
+#| fig-cap: "The frequency distirbution of agents in the training set. "
+#| fig-width: 7
+#| fig-height: 3
+#| out-width: 70%
+hotel_rate_train %>% 
+  count(agent, name = "bookings") %>% 
+  mutate(
+    agent = factor(agent),
+    agent = reorder(agent, bookings)
+  ) %>% 
+  ggplot(aes(bookings)) + 
+  geom_histogram(bins = 40, col = "white") + 
+  scale_x_log10() + 
+  labs(x = "Number of Bookings per Agent")
+```
+
+This section will consider several methods for effectively representing high cardinality predictors. The first two techniques are unsupervised (i.e., they do not use the outcome), while the others are supervised. @Cerda2022 surveys different methods for dealing with these types of predictors. 
+
+### "Othering" {#sec-other}
+
+```{r}
+#| include: false
+
+freq_cut <- 1 / 20
+low_nums <- 
+  hotel_rate_train %>% 
+  count(agent) %>% 
+  mutate(pct = n / nrow(hotel_rate_train) * 100) %>% 
+  filter(pct < freq_cut) %>% 
+  nrow()
+```
+
+It is possible to isolate infrequently occurring categories into a new "other" category. For example, there are `r low_nums` agents with fewer bookings than `r freq_cut`{{< pct >}} of the training set. It may make sense to group these agents into their own categories as long as the predictive ability of the variable is maintained. There is no magic number for the training set frequency cutoff that decides which categories to pool. This pre-processing value requires optimization.
```

Comment:
```suggestion
It is possible to isolate infrequently occurring categories into a new "other" category. For example, there are `r low_nums` agents with fewer bookings than `r freq_cut`{{< pct >}} of the training set. It may make sense to group these agents into their own categories as long as the predictive ability of the variable is maintained. There is no magic number for the training set frequency cutoff that decides which categories to pool. This pre-processing value requires optimization since each data set will have its own nuances with infrequent categories.
```